bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java

```diff
@@ -470,14 +463,68 @@ public void start(BundleContext context) throws Exception {
 		// register debug options listener
 		Hashtable<String, String> properties = new Hashtable<>(2);
 		properties.put(DebugOptions.LISTENER_SYMBOLICNAME, PI_RESOURCES);
-		debugRegistration = context.registerService(DebugOptionsListener.class, Policy.RESOURCES_DEBUG_OPTIONS_LISTENER, properties);
-		// Remember workspace before opening, to
-		// make it easier to debug cases where open() is failing.
-		workspace = new Workspace();
-		IStatus result = workspace.open(null);
-		if (!result.isOK())
-			getLog().log(result);
-		workspaceRegistration = context.registerService(IWorkspace.class, workspace, null);
+		debugRegistration = context.registerService(DebugOptionsListener.class, Policy.RESOURCES_DEBUG_OPTIONS_LISTENER,
+				properties);
+		instanceLocationTracker = new ServiceTracker<>(context,
+				context.createFilter(String.format("(&%s(%s=*))", Location.INSTANCE_FILTER, //$NON-NLS-1$
+						Location.SERVICE_PROPERTY_URL)),
+				new WorkspaceInitCustomizer(context));
+		instanceLocationTracker.open();
+	}
+
+	private final class WorkspaceInitCustomizer implements ServiceTrackerCustomizer<Location, Workspace> {
+		private final BundleContext context;
+		private ServiceRegistration<IWorkspace> workspaceRegistration;
+
+		private WorkspaceInitCustomizer(BundleContext context) {
+			this.context = context;
+		}
+
+		@Override
+		public Workspace addingService(ServiceReference<Location> reference) {
+			if (workspace == null) {
+				Location location = context.getService(reference);
+				if (location != null) {
+					// Remember workspace before opening, to
+					// make it easier to debug cases where open() is failing.
+					workspace = new Workspace();
```

Comment:
there is a race condition possible now.
If the client is in getWorkspace() call at line 431, it can get an unitialized workspace, before it is opened.
You can put breakpoint at line 493 to see what I mean.

Reply:
Fixed that.